### PR TITLE
dc_recovery: optimization: use server-send for small/medium keys

### PR DESCRIPTION
**Pull request by abudnik**
For detailed information see recovery/recovery.qdoc

This optimization speed up in many times dc recovery process for small/medium keys (< chunk-size == 1Mb by default).

### CODE REVIEW

**Comment by agend** [comment]
Is it save to  use? We actually need to add several nodes to our cluster and we can use it and test how it will go. Make sense?

File: recovery/elliptics_recovery/dc_server_send.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+import os
+import errno
+import time
+
+from elliptics_recovery.utils.misc import dump_key_data, load_key_data_from_file
+
+import elliptics
+
+log = logging.getLogger()
+
```
Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: Please add documentation for these new classes and methods, it is really hard to follow this new recovery mechanism without understanding what new classes are needed for

File: recovery/elliptics_recovery/dc_server_send.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    def __init__(self, ctx, node):
+        self.ctx = ctx
+        self.stats = ctx.stats['recover']
+        self.node = node
+        self.buckets = BucketsManager(ctx)
+
+        self.session = elliptics.Session(node)
+        self.session.trace_id = ctx.trace_id
+        self.session.exceptions_policy = elliptics.exceptions_policy.no_exceptions
+        self.session.timeout = 120 # magic ?
```
Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: Why is it magic?

120 seconds should probably be large enough to wait for the first iterator result since it may skip many keys, but in server-send case results should be received very quickly after beginning of the call

Comment (abudnik) [inline recovery/elliptics_recovery/dc_server_send.py]: I've just copied this value from tests/server_send.cpp.
So, is it necessary to set this timeout to 120?

Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: I do not remember default timeout, is it 30 seconds?
This is not needed for `server_send()` method, but may be needed for iterator especially if it requires metadata

Comment (abudnik) [inline recovery/elliptics_recovery/dc_server_send.py]:
```
$> dnet_recovery --help
...
  -w WAIT_TIMEOUT, --wait-timeout=WAIT_TIMEOUT
                        [Wait timeout for elliptics operations default: 3600]
```

I will use default and customizable wait-timeout value.

Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: This was a timeout for the whole operation, such huge values are not required anymore.
Basically, this timeout should reflect time for iterator to send next item, it may take a while if we have ranges - iterator will skip keys and client may reset connection waiting for the next one; and also this timeout should reflect potentially random reads and writes.

**Comment by shaitan** [comment]
@agend This PR includes changes only in dc recovery (replicas synchronization), but you need merge recovery which will redistribute data in the cluster after adding new nodes. Changes in merge recovery is on the way and will be submitted in a separate PR.

**Comment by agend** [comment]
@shaitan Cool! Do you have ETA for it? We can really test it)

**Comment by shaitan** [comment]
@agend I think PR with merge recovery will be submitted next week, so you will can taste it and we would appreciate if you share your experience of using it.

File: recovery/elliptics_recovery/dc_server_send.py (hunk 2 above)
Comment (shaitan) [inline recovery/elliptics_recovery/dc_server_send.py]: Default `wait_timeout` is 5 seconds, 30 seconds - is the default `check_timeout`. `server_send` can use timeout specified by `-w WAIT_TIMEOUT`, but its default should be decreased to some empirical value (`3600` was a upper bound which should never be achieved).

File: recovery/elliptics_recovery/dc_server_send.py (hunk 1 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_server_send.py]: https://github.com/abudnik/elliptics/commit/78a769277baebf59e33bc1bf0f2a3bdaae871a80

**Comment by bioothod** [comment]
Does it need new tests or the old ones are enough?

File: recovery/elliptics_recovery/utils/misc.py [hunk 3]
```diff
@@ -327,6 +327,13 @@ def load_key_data(filepath):
             yield (elliptics.Id(data[0], 0), tuple(KeyInfo.load(d) for d in data[1]))
 
 
+def load_key_data_from_file(keys_file):
```
Comment (shaitan) [inline recovery/elliptics_recovery/utils/misc.py]: It repeats `load_key_data`. Why do you need both of them?

File: recovery/recovery.qdoc [hunk 4]
```diff
@@ -203,7 +207,19 @@
...
-  Built-in implementation of custom recovery for each key from merged results:\n
+  Built-in implementation of custom recovery for each key from merged results consists of two steps: server-send recovery with following rest-keys recovery.\n
+  If key could not be recovered via server-send, then it will be processed during rest-keys recovery.\n
+
+  Server-send recovery:
+  -# while there are non-empty buckets:
+    -# get next bucket - file that contains newest keys from a specific group
+    -# iterate over bucket keys:
+      -# check that key could be recovered via server-send, if so:
+        -# get destination groups to which key must be recovered and call server-send for it
```
Comment (shaitan) [inline recovery/recovery.qdoc]: According to this, recovery calls server-send for each keys. Doesn't it combine keys into batches and call server-send for each batches, does it?

File: recovery/recovery.qdoc [hunk 5]
```diff
@@ -203,7 +207,19 @@
...
 
-  Built-in implementation of custom recovery for each key from merged results:\n
+  Built-in implementation of custom recovery for each key from merged results consists of two steps: server-send recovery with following rest-keys recovery.\n
+  If key could not be recovered via server-send, then it will be processed during rest-keys recovery.\n
+
+  Server-send recovery:
+  -# while there are non-empty buckets:
+    -# get next bucket - file that contains newest keys from a specific group
+    -# iterate over bucket keys:
+      -# check that key could be recovered via server-send, if so:
```
Comment (shaitan) [inline recovery/recovery.qdoc]: It should be described how recovery checks this.

File: recovery/recovery.qdoc [hunk 6]
```diff
@@ -203,7 +207,19 @@
...
+        -# get destination groups to which key must be recovered and call server-send for it
+        -# check server-send result for the key: if operation failed, then move key to the other bucket if possible, otherwise to 'rest-keys'
```
Comment (shaitan) [inline recovery/recovery.qdoc]: Can you describe how a key migrates between different buckets including 'rest_keys' bucket.

File: recovery/recovery.qdoc [hunk 7]
```diff
@@ -195,6 +195,10 @@
   -# goes throw merged results and dumps all iterated keys.
     This dump file could be used for resuming recovery after some issues.
     Dump file has name 'dump' and is placed in tmp_dir specified by <b>-D DIR, --dir=DIR</b>.
+  -# distributes keys among multiple files (buckets).
```
Comment (shaitan) [inline recovery/recovery.qdoc]:
```
-# distributes keys among buckets:
    one bucket with name 'rest_keys' will contain keys which can't be recovered via server-send
    number of group buckets with name 'bucket_xx' ('xx' is group_id) will contain keys
    which can be recovered via server-send from corresponding group.
```

File: recovery/elliptics_recovery/types/dc.py [hunk 8]
```diff
@@ -314,6 +371,10 @@ def lookup_keys(ctx):
     session = elliptics.Session(node)
     session.trace_id = ctx.trace_id
     filename = os.path.join(ctx.tmp_dir, 'merged_result')
+    ctx.rest_keys_filename = os.path.join(ctx.tmp_dir, 'rest_keys')
+    ctx.rest_file = open(ctx.rest_keys_filename, 'wb')
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: `ctx.rest_keys_filename` can be get via `cx.rest_file.name`.

File: recovery/elliptics_recovery/types/dc.py [hunk 9]
```diff
@@ -222,23 +227,71 @@ def contains(addresses_with_backends, address, backend_id):
...
-                os.remove(res[1])
-
-    log.debug("Merged_filename: %s, address: %s, groups: %s, tmp_dir: %s",
-              ctx.merged_filename, ctx.address, ctx.groups, ctx.tmp_dir)
+def dump_key(ctx, rest_file, key, key_infos, newest_key_group):
+    if key_infos[0].group_id != newest_key_group:
+        index = [k.group_id for k in key_infos].index(newest_key_group)
+        tmp = list(key_infos)
+        tmp[0], tmp[index] = tmp[index], tmp[0]
+        key_infos = tuple(tmp)
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: What is the logic behind the code above?

File: recovery/elliptics_recovery/types/dc.py [hunk 10]
```diff
@@ -340,11 +401,21 @@ def lookup_keys(ctx):
                         stats.counter("lookups", -1)
                 if len(key_infos) > 0:
                     key_data = (id, key_infos)
-                    dump_key_data(key_data, merged_f)
+                    if not skip_key_data(ctx, key_data):
+                        key_infos.sort(key=lambda x: (x.timestamp, x.size), reverse=True)
+                        newest_key_group = key_infos[0].group_id
+                        dump_key_data(key_data, merged_f)
+                        dump_key(ctx, ctx.rest_file, id, key_infos, newest_key_group)
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: Why does it dump keys twice?

File: recovery/elliptics_recovery/types/dc.py [hunk 11]
```diff
@@ -222,23 +227,71 @@ def contains(addresses_with_backends, address, backend_id):
...
-        with open(dump_filename, 'wb') as df:
-            for res in (r for r in results if r):
-                shutil.copyfileobj(open(res[0], 'rb'), mf)
-                os.remove(res[0])
-                shutil.copyfileobj(open(res[1], 'rb'), df)
-                os.remove(res[1])
-
-    log.debug("Merged_filename: %s, address: %s, groups: %s, tmp_dir: %s",
-              ctx.merged_filename, ctx.address, ctx.groups, ctx.tmp_dir)
+def dump_key(ctx, rest_file, key, key_infos, newest_key_group):
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: It has `ctx`, why it doesn't use `ctx.rest_file` instead of `rest_file`?

File: recovery/elliptics_recovery/types/dc.py [hunk 12]
```diff
@@ -222,23 +227,71 @@ def contains(addresses_with_backends, address, backend_id):
...
+One bucket is 'rest_keys' and other buckets are 'bucket_xx', where xx == group_id.
+'bucket_xx' contains newest keys that should be recovered from group xx to other groups
+via server_send. If a key could not be recovered with server_send it is placed to 'rest_keys'.
+
+Also this function prepares bucket_order array. The array contains group_id's sorted by amount
+of keys in appropriate bucket. This array will be used by server_send recovery process.
+'''
+def fill_buckets(ctx, results):
+    newest_key_stats = dict()
+    for _, _, range_stats in (r for r in results if r):
```
Comment (shaitan) [inline recovery/elliptics_recovery/types/dc.py]: As I see, wherever `results` are used they are filtered via `(r for r in results if r)`. It can be done ones [here] (https://github.com/reverbrain/elliptics/pull/668/files#diff-302f790c9667058c7ff408b010833a2fR328).

File: recovery/elliptics_recovery/dc_recovery.py [hunk 13]
```diff
@@ -402,6 +404,12 @@ def run_one(self):
         return False
 
 
+def cleanup(ctx):
+    for f in ctx.bucket_files.itervalues():
+        os.remove(f.name)
```
Comment (shaitan) [inline recovery/elliptics_recovery/dc_recovery.py]: Why doesn't it close `f`?

File: recovery/elliptics_recovery/dc_recovery.py [hunk 14]
```diff
@@ -412,7 +420,9 @@ def recover(ctx):
                                  net_thread_num=4,
                                  io_thread_num=24,
                                  remotes=ctx.remotes)
+    ServerSendRecovery(ctx, node).recover()
```
Comment (shaitan) [inline recovery/elliptics_recovery/dc_recovery.py]: Why doesn't result of `ServerSendRecovery().recover()` affect to final result of whole recovery? What if it is failed and nothing can be recovered by `WindowDC()`.

File: recovery/elliptics_recovery/dc_recovery.py (hunk 13 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_recovery.py]: ctx.rest_file used in WindowDC

**Comment by agend** [comment]
Guys, do you have plan for this feature?

**Comment by shaitan** [comment]
@abudnik is making changes to merge, so latter this week there will be another PR for merge recovery. After that both PRs will be reviewed and tested.

**Comment by agend** [comment]
@shaitan Cool! Thanks!

**Comment by abudnik** [comment]
git push forced here: this branch was forked from https://github.com/abudnik/elliptics/tree/recovery_ell - common changes in elliptics for both new dc (this PR) and coming merge (PR in process) recoveries.

**Comment by bioothod** [comment]
There are comments by @shaitan which are not yet addressed. It would be great to have more comments about design of the new classes, the logic behind them and so on.
It should help understanding the whole picture of the new recovery process.

**Comment by bioothod** [comment]
@abudnik there are some comments on this pull request, please resolve them and I will merge this pull request

File: recovery/elliptics_recovery/dc_recovery.py (hunk 14 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_recovery.py]: Fixed

File: recovery/elliptics_recovery/dc_recovery.py (hunk 13 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_recovery.py]: Fixed

File: recovery/elliptics_recovery/types/dc.py (hunk 12 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Fixed

File: recovery/elliptics_recovery/types/dc.py (hunk 11 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Fixed, it uses ctx.rest_file

File: recovery/elliptics_recovery/types/dc.py (hunk 10 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Fixed

File: recovery/elliptics_recovery/utils/misc.py (hunk 3 above)
Comment (abudnik) [inline recovery/elliptics_recovery/utils/misc.py]: Fixed

File: recovery/elliptics_recovery/types/dc.py (hunk 8 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: Fixed. I removed rest_keys_filename from ctx.

**Comment by abudnik** [comment]
@bioothod: >Does it need new tests or the old ones are enough?
It seems that old tests cover most important cases.

File: recovery/elliptics_recovery/types/dc.py (hunk 9 above)
Comment (abudnik) [inline recovery/elliptics_recovery/types/dc.py]: First key_info in key_infos list must correspond to largest bucket among other buckets. `newest_key_group` is a group_id of the largest bucket.

File: recovery/recovery.qdoc (hunk 7 above)
Comment (abudnik) [inline recovery/recovery.qdoc]: Fixed

File: recovery/recovery.qdoc (hunk 6 above)
Comment (abudnik) [inline recovery/recovery.qdoc]: Done

File: recovery/recovery.qdoc (hunk 5 above)
Comment (abudnik) [inline recovery/recovery.qdoc]: Done

File: recovery/recovery.qdoc (hunk 4 above)
Comment (abudnik) [inline recovery/recovery.qdoc]: Fixed: `-# iterate over bucket keys by batch:`

File: recovery/elliptics_recovery/dc_server_send.py [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+                    if corrupted_keys:
+                        self._remove_corrupted_keys(corrupted_keys, [group])
+
+            if timeouted_keys:
+                self._on_server_send_timeout(timeouted_keys, key_infos_map, group)
+
+    def _check_server_send_results(self, iterator, key_infos_map, group):
+        '''
+        Check result of remote sending for every key.
+        Returns list of timeouted keys.
```
Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: Why do you call them timeouted keys? It looks like it returns a list of failed keys no matter what was the reason

File: recovery/elliptics_recovery/dc_server_send.py (hunk 1 above)
Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: This is not a documentation, this is a one-liner which says the purpose of the function without description of the logic and idea behind it. Its like saying 'this is a table' pointing your finger to table. You do not say that it is a wood table or that it is needed to host plates or books or why did you bought it and put it exactly here.

In particular, bucket/server_send classes completely lack description of the recovery process when server sending fails. This is just one note among others related to documentation. Please write not only what particular entity does, but what did you want it to do, how you want it to behave and why is this needed.

File: recovery/elliptics_recovery/dc_server_send.py [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+            for k in keys:
+                result = self.remove_session.remove(k)
+                results.append(result)
+
+            timeouted_keys = []
+            timeouted_groups = set()
+            is_last_attempt = (attempt == self.ctx.attempts - 1)
+            for i, r in enumerate(results):
+                status = r.get()[0].status
+                log.info("Removing corrupted key: {0}, status: {1}, last attempt: {2}".format(keys[i], status, is_last_attempt))
```
Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: Do you blindly remove keys with invalid checksum even if there are no other better replicas? Should it recover these keys by reading them without checksum verification?

I do have keys which seem like broken, but they are actually correct, but for some reason they do not have checksum footer and do not have no-checksum bit in their header. I do not want them to be removed.

File: recovery/elliptics_recovery/types/dc.py [hunk 17]
```diff
@@ -225,18 +230,80 @@ def final_merge(ctx, results):
...
+
+def fill_buckets(ctx, results):
+    '''
+    This function distributes keys among multiple files (buckets).
+    One bucket is 'rest_keys' and other buckets are 'bucket_xx', where xx == group_id.
+    'bucket_xx' contains newest keys that should be recovered from group xx to other groups
+    via server_send. If a key could not be recovered with server_send it is placed to 'rest_keys'.
+
+    Also this function prepares bucket_order array. The array contains group_id's sorted by amount
+    of keys in appropriate bucket. This array will be used by server_send recovery process.
```
Comment (bioothod) [inline recovery/elliptics_recovery/types/dc.py]: This is example of good documentation in this pull request.

File: recovery/elliptics_recovery/dc_server_send.py [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+        if len(key_infos) > 1:
+            return key_infos[1].group_id
+        return -1
+
+    def _update_stats(self, start_time, processed_keys, recovers_in_progress, status):
+        speed = processed_keys / (time.time() - start_time)
+        recovers_in_progress -= processed_keys
+        self.stats.set_counter('recovery_speed', round(speed, 2))
+        self.stats.set_counter('recovers_in_progress', recovers_in_progress)
+        self.stats.counter('recovered_keys', 1 if status == 0 else -1)
```
Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: What do these new fields mean? Should the be described in qdoc?

File: recovery/elliptics_recovery/dc_server_send.py (hunk 15 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_server_send.py]: Timeouted and corrupted lists are filled in _on_server_send_fail().

File: recovery/elliptics_recovery/dc_server_send.py (hunk 16 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_server_send.py]: You can use --safe flag. From help message:
```
  -S, --safe            Do not remove recovered keys after merge [default:
                        False]
```

File: recovery/elliptics_recovery/dc_server_send.py (hunk 1 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_server_send.py]: See docs in recovery/recovery.qdoc ('Server-send recovery: ...')

Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: This doesn't tell me how `on_server_send_fail()` behaves in bucket manager and server-send recovery classes and how they are related to each other.

Documentation in the code tells reader how particular **code** is supposed to be designed and how it is supposed to be working. Text is qdoc you reference above tells higher-level user how recovery works.

File: recovery/elliptics_recovery/dc_server_send.py (hunk 16 above)
Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: What about recovering such keys?
With this option, keys will not be read, but they also will not be recovered, is that correct?

File: recovery/elliptics_recovery/dc_server_send.py (hunk 15 above)
Comment (bioothod) [inline recovery/elliptics_recovery/dc_server_send.py]: There is also `buckets.on_server_send_fail()` call, which does something with the key. Will it change the key? Or put into some other place?

Please describe how these new classes should work, what is the design decision behind this code, it is vital knowledge to understand and fix bugs.

File: recovery/elliptics_recovery/dc_server_send.py (hunk 16 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_server_send.py]: With this option, keys will not be removed.
Keys with invalid checksum will not be recovered via server-send, because they have invalid checksum.

File: recovery/elliptics_recovery/dc_server_send.py (hunk 1 above)
Comment (abudnik) [inline recovery/elliptics_recovery/dc_server_send.py]: Added more comments on handling server-send (recovery) failures: https://github.com/abudnik/elliptics/commit/cbbad2448b5a2da2d7e86b41b364f38edb51ae1d

**Comment by abudnik** [comment]
This PR is ready to be merged.
